crates/bevy_animation/src/gltf_curves.rs

```diff
@@ -0,0 +1,420 @@
+    ///   filtered, and deduplicated).
+    #[inline]
+    pub fn new(
+        times: impl IntoIterator<Item = f32>,
+        values: impl IntoIterator<Item = T>,
+    ) -> Result<Self, WideKeyframeCurveError> {
+        let times: Vec<f32> = times.into_iter().collect();
+        let values: Vec<T> = values.into_iter().collect();
+        let divisor = times.len() * 3;
+
+        if values.len() % divisor != 0 {
+            return Err(WideKeyframeCurveError::LengthMismatch {
+                values_given: values.len(),
+                divisor,
+            });
+        }
+
+        Ok(Self {
+            core: ChunkedUnevenCore::new_width_inferred(times, values)?,
+        })
+    }
+}
+
+/// A curve specifying the [`MorphWeights`] for a mesh in animation. The variants are broken
+/// down by interpolation mode (with the exception of `Constant`, which never interpolates).
+///
+/// This type is, itself, a `Curve<Vec<f32>>`; however, in order to avoid allocation, it is
+/// recommended to use its implementation of the [`IterableCurve`] trait, which allows iterating
+/// directly over information derived from the curve without allocating.
+///
+/// [`MorphWeights`]: bevy_render::prelude::MorphWeights
+#[derive(Debug, Clone, Reflect)]
+pub enum WeightsCurve {
+    /// A curve which takes a constant value over its domain. Notably, this is how animations with
+    /// only a single keyframe are interpreted.
+    Constant(ConstantCurve<Vec<f32>>),
+
+    /// A curve which interpolates weights linearly between keyframes.
+    Linear(WideLinearKeyframeCurve<f32>),
+
+    /// A curve which interpolates weights between keyframes in steps.
+    Step(WideSteppedKeyframeCurve<f32>),
+
+    /// A curve which interpolates between keyframes by using auxiliary tangent data to join
+    /// adjacent keyframes with a cubic Hermite spline, which is then sampled.
+    CubicSpline(WideCubicKeyframeCurve<f32>),
+}
+
+//---------//
+// HELPERS //
+//---------//
+
+enum TwoIterators<A, B> {
+    Left(A),
+    Right(B),
+}
+
+impl<A, B, T> Iterator for TwoIterators<A, B>
+where
+    A: Iterator<Item = T>,
+    B: Iterator<Item = T>,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            TwoIterators::Left(a) => a.next(),
+            TwoIterators::Right(b) => b.next(),
+        }
+    }
+}
+
+/// Helper function for cubic spline interpolation.
+fn cubic_spline_interpolation<T>(
+    value_start: T,
+    tangent_out_start: T,
+    tangent_in_end: T,
+    value_end: T,
+    lerp: f32,
+    step_duration: f32,
+) -> T
+where
+    T: VectorSpace,
+{
+    value_start * (2.0 * lerp.cubed() - 3.0 * lerp.squared() + 1.0)
+        + tangent_out_start * (step_duration) * (lerp.cubed() - 2.0 * lerp.squared() + lerp)
+        + value_end * (-2.0 * lerp.cubed() + 3.0 * lerp.squared())
+        + tangent_in_end * step_duration * (lerp.cubed() - lerp.squared())
```

Comment:
I suspect this would be clearer, more numerically stable, and possibly faster with de Casteljau subdivision (repeated nested lerps), the way I had it before. At the very least this should be using [Horner's method](http://en.wikipedia.org/wiki/Horner's_method) to evaluate the polynomials faster.

Reply:
If this method was actually called often (with the same coefficients), what you should do is pre-compute the vectors A,B,C,D, such that this method equals `A + tB + t^2 C + t^3 D`, and then eval this polynomial with Horner's method. (Like we do in the CubicSegment.)

Reply:
Yeah, this being the old version is basically a merge artifact, since I moved the old code verbatim out of `lib` into what is now `gltf_curves`, while you moved it into `animatable` and changed it. I don't have any attachment to the old version at all, so we could just axe it if we're confident the new one is better.

Reply:
Here's a proposal:
```rust
let coeffs = (vec4(2.0, 1.0, -2.0, 1.0) * lerp + vec4(-3.0, -2.0, 3.0, -1.0) * lerp;
value_start * (coeffs.x + 1.0)
+ tangent_out_start * step_duration * lerp * coeffs.y
+ value_end * coeffs.z
+ tangent_in_end * coeffs.w
```

I doubt you can go much faster than this.

Reply:
I think it should be like this, if I'm not mistaken (I think it was missing a couple lerps and a +1):
```rust
let coeffs = (vec4(2.0, 1.0, -2.0, 1.0) * lerp + vec4(-3.0, -2.0, 3.0, -1.0)) * lerp;
value_start * (coeffs.x * lerp + 1.0)
    + tangent_out_start * step_duration * lerp * (coeffs.y + 1.0)
    + value_end * lerp * coeffs.z
    + tangent_in_end * step_duration * lerp * coeffs.w
```

Reply:
Oh right, I see.

Yeah this is most likely going to be a bunch faster. I did not realise you meant we can batch the coefficients like this (which should hopefully use SIMD under the hood too.)